ppafm/ocl/field.py

```diff
@@ -241,10 +244,31 @@ def power_positive(self, p=1.2, in_place=True, local_size=(32,), queue=None):
             grid_out = array_type(array_out, lvec=self.lvec, shape=self.shape, ctx=self.ctx)
         n = np.int32(array_in.size / 4)
         p = np.float32(p)
+        if normalize:
+            scale = np.float32(self._get_normalization_factor(queue)) ** p
+        else:
+            scale = np.float32(1.0)
         global_size = [int(np.ceil(n / local_size[0]) * local_size[0])]
-        cl_program.power(queue, global_size, local_size, array_in, array_out, n, p)
+        cl_program.power(queue, global_size, local_size, array_in, array_out, n, p, scale)
         return grid_out
 
+    def _get_normalization_factor(self, queue=None):
+        queue = queue or oclu.queue
+        n = np.int32(np.prod(self.shape))
+        local_size = (256,)
+        n_groups = np.int32(min(local_size[0], (n - 1) // local_size[0] + 1))
+        global_size = (local_size[0] * n_groups,)
+        array_in = self.cl_array
+        array_out = cl.Buffer(self.ctx, cl.mem_flags.READ_WRITE, size=8*n_groups)
+        # First do sums of the input array within each work group...
+        cl_program.normalizeSumReduce(queue, global_size, local_size, array_in, array_out, n)
+        # ... then sum the results of the first kernel call
+        cl_program.sumSingleGroup(queue, local_size, local_size, array_out, n_groups)
```

Comment:
Yes, I purposefully set the number of groups in the first pass to be at most the size of the work group so that in the second pass it can be done by a single work group.